### PR TITLE
feat: search index persistence

### PR DESCRIPTION
**Pull request by dranikpg**
This PR adds persistence for search indices

Their definitions are stored in plain string form in AUX fields

### CODE REVIEW

File: src/server/rdb_load.cc [hunk 1]
```diff
@@ -2339,4 +2333,49 @@ error_code RdbLoader::LoadKeyValPair(int type, ObjSettings* settings) {
...
+  facade::RespVec resp_vec;
+  facade::RedisParser parser;
+
+  absl::Span<uint8_t> buffer{reinterpret_cast<uint8_t*>(def.data()), def.size()};
+  auto res = parser.Parse(buffer, &consumed, &resp_vec);
+
+  if (res != facade::RedisParser::Result::OK) {
+    LOG(ERROR) << "Bad index definition: " << def;
+    return;
+  }
```
Comment (dranikpg) [inline src/server/rdb_load.cc:2365]: Script compilation (see above) fails silently, so this should too?

File: src/server/rdb_load.cc [hunk 2]
```diff
@@ -2339,4 +2333,49 @@ error_code RdbLoader::LoadKeyValPair(int type, ObjSettings* settings) {
...
+
+  absl::Cleanup cntx_clean = [&cntx] { cntx.Inject(nullptr); };
+
+  uint32_t consumed = 0;
+  facade::RespVec resp_vec;
+  facade::RedisParser parser;
+
+  absl::Span<uint8_t> buffer{reinterpret_cast<uint8_t*>(def.data()), def.size()};
+  auto res = parser.Parse(buffer, &consumed, &resp_vec);
+
```
Comment (dranikpg) [inline src/server/rdb_load.cc:2361]: Technically we don't need this bulky parser code, we can just split by whitespace (taking into account we build the string like this without any formatting). So RESP would be needed only for quotation marks? 🤔

**Comment by kostasrim** [comment]
Nice work :man_cook:

File: src/server/rdb_save.cc [hunk 3]
```diff
@@ -1149,15 +1149,15 @@ void RdbSaver::StopSnapshotInShard(EngineShard* shard) {
...
+error_code RdbSaver::SaveHeader(const GlobalData& glob_state) {
   char magic[16];
   // We should use RDB_VERSION here from rdb.h when we ditch redis 6 support
   // For now we serialize to an older version.
   size_t sz = absl::SNPrintF(magic, sizeof(magic), "REDIS%04d", RDB_SER_VERSION);
   CHECK_EQ(9u, sz);
 
   RETURN_ON_ERR(impl_->serializer()->WriteRaw(Bytes{reinterpret_cast<uint8_t*>(magic), sz}));
-  RETURN_ON_ERR(SaveAux(lua_scripts));
+  RETURN_ON_ERR(SaveAux(move(glob_state)));
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/rdb_save.cc:1160]: True, will use fully qualified move

File: src/server/rdb_save.cc [hunk 4]
```diff
@@ -1197,16 +1197,18 @@ error_code RdbSaver::SaveAux(const StringVec& lua_scripts) {
...
 
   // Save lua scripts only in rdb or summary file
-  DCHECK(save_mode_ != SaveMode::SINGLE_SHARD || lua_scripts.empty());
-  for (const string& s : lua_scripts) {
+  DCHECK(save_mode_ != SaveMode::SINGLE_SHARD || glob_state.lua_scripts.empty());
+  for (const string& s : glob_state.lua_scripts)
     RETURN_ON_ERR(impl_->SaveAuxFieldStrStr("lua", s));
-  }
+
+  // Save search index definitions only in summary dfs file.
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/rdb_save.cc:1208]: Yes, your points are correct, I intend to save only inside the dfs summary

File: src/server/rdb_save.cc [hunk 5]
```diff
@@ -1197,15 +1197,22 @@ error_code RdbSaver::SaveAux(const StringVec& lua_scripts) {
...
   RETURN_ON_ERR(SaveAuxFieldStrInt("aof-preamble", aof_preamble));
 
   // Save lua scripts only in rdb or summary file
-  DCHECK(save_mode_ != SaveMode::SINGLE_SHARD || lua_scripts.empty());
-  for (const string& s : lua_scripts) {
+  DCHECK(save_mode_ != SaveMode::SINGLE_SHARD || glob_state.lua_scripts.empty());
+  for (const string& s : glob_state.lua_scripts)
     RETURN_ON_ERR(impl_->SaveAuxFieldStrStr("lua", s));
+
+  if (save_mode_ == SaveMode::RDB) {
```
Comment (kostasrim) [inline src/server/rdb_save.cc:1208]: Nice! So that was a bug!

**Review by kostasrim — APPROVED** [review]

File: src/server/rdb_save.cc (hunk 5 above)
Comment (dranikpg) [reply src/server/rdb_save.cc:1208]: No, I just print a warnings now 😄 Unknown aux fields are ignored either way

**Pull request merged**